src/vm/i386/asmhelpers.asm

```diff
@@ -1595,6 +1598,51 @@ public _ExternalMethodFixupPatchLabel@0
 
 _ExternalMethodFixupStub@0 endp
 
+;==========================================================================
+; This is the stub called from ZapExternalMethodCell indirection cell in native images.
+; It sets up parameters to the ExternalMethodFixupWorker so that pIndirection
+; points to the indirection cell, and thus the worker will patch the indirection cell.
+_ExternalMethodCellFixupStub@0 proc public
+
+    mov         eax, [esp]  ; remeber the caller's return address
+                            ; used later to figure out the indirection cell address
+                            ; which was used to call this stub.
+
+    STUB_PROLOG
+
+    mov         esi, esp
+
+    ; EAX is return address into the callsite which has "call [cell]"
+    ; subtract 4 to get the address of the call target, which points to the indirection cell.
+    ; From there load the address of the indirection cell.
+    mov         eax, [eax - 4]
```

Comment:
I think if you kept this as null, it would work as well. The start of ExternalMethodFixupWorker has similar code that disasembles the callsite when the indirection cell is null.

Reply:
I actually tried that, but the code in the worker is broken. More specifically the ExternalMethodFrame is broken (the logic in it to compute the return address is wrong). I didn't fully understand what the frames are for, so I picked the solution with the stub doing this. To me it feels cleaner, as we have separate files for each architecture and can potentially even delete the ifdef mess at the beginning of the worker.

Reply:
> the logic in it to compute the return address is wrong

This sounds suspect. If ExternalMethodFrame gets its return address wrong, the stackwalking through it will be broken too - GC will crash if it happens to run when ExternalMethodFrame is on the stack. It may be worth understanding how it is wrong.